src/rgw/rgw_bucket.cc

```diff
@@ -1411,6 +1413,116 @@ static int bucket_stats(RGWRados *store, const std::string& tenant_name, std::st
   return 0;
 }
 
+int RGWBucketAdminOp::limit_check(RGWRados *store,
+				  RGWBucketAdminOpState& op_state,
+				  const std::list<std::string>& user_ids,
+				  RGWFormatterFlusher& flusher)
+{
+  int ret = 0;
+  const size_t max_entries =
+    store->ctx()->_conf->rgw_list_buckets_max_chunk;
+
+  const size_t safe_max_objects_per_shard =
+    store->ctx()->_conf->rgw_safe_max_objects_per_shard;
+
+  uint16_t shard_warn_pct =
+    store->ctx()->_conf->rgw_shard_warning_threshold;
+  if (shard_warn_pct > 100)
+    shard_warn_pct = 10;
+
+  Formatter *formatter = flusher.get_formatter();
+  flusher.start(0);
+
+  formatter->open_array_section("users");
+
+  for (const auto& user_id : user_ids) {
+    formatter->open_object_section("user");
+    formatter->dump_string("user_id", user_id);
+    bool done;
+    formatter->open_array_section("buckets");
+    do {
+      RGWUserBuckets buckets;
+      string marker;
+      bool is_truncated;
+
+      ret = rgw_read_user_buckets(store, user_id, buckets,
+				  marker, string(), max_entries, false,
+				  &is_truncated);
+      if (ret < 0)
+        return ret;
+
+      map<string, RGWBucketEnt>& m_buckets = buckets.get_buckets();
+
+      for (const auto& iter : m_buckets) {
+	auto& bucket = const_cast<rgw_bucket&>(iter.second.bucket);
+	uint32_t num_shards = 1;
+	uint64_t num_objects = 0;
+
+	/* need stats for num_entries */
+	string bucket_ver, master_ver;
+	std::map<RGWObjCategory, RGWStorageStats> stats;
+	ret = store->get_bucket_stats(bucket, RGW_NO_SHARD, &bucket_ver,
+				      &master_ver, stats, nullptr);
+	if (! ret) {
+	  for (const auto& s : stats) {
+	    num_objects = s.second.num_objects;
+	  }
+	}
+
+	/* need info for num_shards */
+	RGWBucketInfo bucket_info;
+	RGWObjectCtx obj_ctx(store);
+
+	ret = store->get_bucket_info(obj_ctx, bucket.tenant, bucket.name,
+				     bucket_info, nullptr);
+	if (! ret) {
+	  if (bucket_info.num_shards > 0)
+	    num_shards = bucket_info.num_shards;
+	}
+
+	uint64_t objs_per_shard = num_objects / num_shards;
+
+	formatter->open_object_section("bucket");
+	formatter->dump_string("bucket", bucket.name);
+	formatter->dump_string("tenant", bucket.tenant);
+	formatter->dump_int("num_objects", num_objects);
+	formatter->dump_int("num_shards", num_shards);
+	formatter->dump_int("objects_per_shard", objs_per_shard);
+	{
+	  stringstream ss;
+	  if (objs_per_shard > safe_max_objects_per_shard) {
+	    ss << boost::format("OVER %4f%%") %
+	      (safe_max_objects_per_shard/objs_per_shard * 100);
+	  } else {
+	    double safe_fill_pct =
+	      100 - (objs_per_shard / safe_max_objects_per_shard * 100);
+	    if (safe_fill_pct <= shard_warn_pct) {
+	      ss << boost::format("WARN %4f%%") % safe_fill_pct;
+	    } else {
+	      ss << "OK";
```

Comment:
@mattbenjamin maybe we need to have two modes of operations? one that just dumps everything (like here), another one that just dumps warnings and errors?